frame/support/src/lib.rs

```diff
@@ -65,32 +142,46 @@ pub mod structs {
 
 	#[derive(Clone, Default, PartialEq, Eq, Encode, Decode, RuntimeDebug)]
 	pub struct StakingLock<Balance, Moment> {
+		/// The amount which the free balance may not drop below when this lock is in effect.
 		pub staking_amount: Balance,
 		pub unbondings: Vec<NormalLock<Balance, Moment>>,
 	}
 
 	impl<Balance, Moment> StakingLock<Balance, Moment>
 	where
-		Balance: Copy + PartialOrd + SimpleArithmetic,
+		Balance: Copy + PartialOrd + AtLeast32Bit,
 		Moment: Copy + PartialOrd,
 	{
+		#[inline]
+		pub fn locked_amount(&self, at: Moment) -> Balance {
+			self.unbondings.iter().fold(self.staking_amount, |acc, unbonding| {
+				if unbonding.valid_at(at) {
+					acc + unbonding.amount
+				} else {
+					acc
+				}
+			})
+		}
+
 		#[inline]
 		fn can_withdraw(&self, at: Moment, new_balance: Balance) -> bool {
+			new_balance >= self.locked_amount(at)
+		}
+
+		#[inline]
+		fn update_locks(&mut self, at: Moment) -> Balance {
```

Comment:
This is only called when set_lock(), suggest to move to staking model, to keep lock logic simple.